File: client/src/test/java/org/octopusden/octopus/jira/api/client/impl/ClassicJiraApiClientTest.java
package org.octopusden.octopus.jira.api.client.impl;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.Before;
import org.junit.Test;
import org.octopusden.octopus.jira.api.config.JacksonMapper;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Base64;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ClassicJiraApiClientTest {

    private ObjectMapper objectMapper;
    private JiraApiClientParametersProvider parametersProvider;

    @Before
    public void setup() {
        objectMapper = JacksonMapper.create();
        parametersProvider = mock(JiraApiClientParametersProvider.class);
        when(parametersProvider.getApiUrl()).thenReturn("http://localhost:8080");
    }

    @Test
    public void getAuthHeaderReturnsBearerTokenWhenTokenIsPresent() {
        when(parametersProvider.getBearerToken()).thenReturn("my-token");
        ClassicJiraApiClient client = new ClassicJiraApiClient(parametersProvider, objectMapper);

        String authHeader = invokeGetAuthHeader(client);

        assertEquals("Bearer my-token", authHeader);
    }

    @Test
    public void getAuthHeaderReturnsBasicCredentialsWhenTokenIsNull() {
        when(parametersProvider.getBearerToken()).thenReturn(null);
        when(parametersProvider.getBasicCredentials()).thenReturn("user:pass");
        ClassicJiraApiClient client = new ClassicJiraApiClient(parametersProvider, objectMapper);

        String authHeader = invokeGetAuthHeader(client);

        assertTrue(authHeader != null && authHeader.startsWith("Basic "));
        byte[] decoded = Base64.getDecoder().decode(authHeader.substring("Basic ".length()));
        assertEquals("user:pass", new String(decoded));
    }

    @Test
    public void bearerTokenTakesPrecedenceOverBasicCredentials() {
        when(parametersProvider.getBearerToken()).thenReturn("bearer-token");
        when(parametersProvider.getBasicCredentials()).thenReturn("user:pass");
        ClassicJiraApiClient client = new ClassicJiraApiClient(parametersProvider, objectMapper);

        String authHeader = invokeGetAuthHeader(client);

        assertEquals("Bearer bearer-token", authHeader);
    }

    @Test
    public void getAuthHeaderWhenBothTokenAndCredentialsAreNull() {
        when(parametersProvider.getBearerToken()).thenReturn(null);
        when(parametersProvider.getBasicCredentials()).thenReturn(null);
        ClassicJiraApiClient client = new ClassicJiraApiClient(parametersProvider, objectMapper);

        assertNull(invokeGetAuthHeader(client));
    }

    @Test
    public void getAuthHeaderWhenBothAreBlank() {
        when(parametersProvider.getBearerToken()).thenReturn("");
        when(parametersProvider.getBasicCredentials()).thenReturn(":");
        ClassicJiraApiClient client = new ClassicJiraApiClient(parametersProvider, objectMapper);

        assertNull(invokeGetAuthHeader(client));
    }

    private String invokeGetAuthHeader(ClassicJiraApiClient client) {
        try {
            Method method = ClassicJiraApiClient.class.getDeclaredMethod("getAuthHeader");
            method.setAccessible(true);
            return (String) method.invoke(client);
        } catch (InvocationTargetException e) {
            Throwable cause = e.getCause();
            if (cause instanceof RuntimeException) throw (RuntimeException) cause;
            throw new RuntimeException(cause);
        } catch (ReflectiveOperationException e) {
            throw new RuntimeException(e);
        }
    }
}

File: client/src/test/kotlin/org/octopusden/octopus/jira/api/client/impl/ClassicJiraApiClientTest.kt
package org.octopusden.octopus.jira.api.client.impl

import com.fasterxml.jackson.databind.ObjectMapper
import io.mockk.every
import io.mockk.mockk
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.octopusden.octopus.jira.api.config.JacksonMapper
import kotlin.test.assertNull

class ClassicJiraApiClientTest {

    private lateinit var objectMapper: ObjectMapper
    private lateinit var parametersProvider: JiraApiClientParametersProvider

    @Before
    fun setup() {
        objectMapper = JacksonMapper.create()
        parametersProvider = mockk(relaxed = true)
        every { parametersProvider.getApiUrl() } returns "http://localhost:8080"
    }

    @Test
    fun `getAuthHeader returns Bearer token when token is present`() {
        every { parametersProvider.getBearerToken() } returns "my-token"
        val client = ClassicJiraApiClient(parametersProvider, objectMapper)

        val authHeader = invokeGetAuthHeader(client)

        assertEquals("Bearer my-token", authHeader)
    }

    @Test
    fun `getAuthHeader returns Basic credentials when token is null`() {
        every { parametersProvider.getBearerToken() } returns null
        every { parametersProvider.getBasicCredentials() } returns "user:pass"
        val client = ClassicJiraApiClient(parametersProvider, objectMapper)

        val authHeader = invokeGetAuthHeader(client)!!

        assertTrue(authHeader.startsWith("Basic "))
        val decoded = java.util.Base64.getDecoder().decode(authHeader.removePrefix("Basic "))
        assertEquals("user:pass", String(decoded))
    }

    @Test
    fun `Bearer token takes precedence over basic credentials`() {
        every { parametersProvider.getBearerToken() } returns "bearer-token"
        every { parametersProvider.getBasicCredentials() } returns "user:pass"
        val client = ClassicJiraApiClient(parametersProvider, objectMapper)

        val authHeader = invokeGetAuthHeader(client)

        assertEquals("Bearer bearer-token", authHeader)
    }

    @Test
    fun `getAuthHeader when both token and credentials are null`() {
        // Stub to allow constructor to succeed (Feign builder doesn't trigger interceptor until first request),
        // then set both to null before calling getAuthHeader via reflection.
        every { parametersProvider.getBearerToken() } returns "temp"
        val client = ClassicJiraApiClient(parametersProvider, objectMapper)
        // Now override to null — the Feign interceptor won't fire until an actual HTTP call
        every { parametersProvider.getBearerToken() } returns null
        every { parametersProvider.getBasicCredentials() } returns null
        assertNull(invokeGetAuthHeader(client))
    }

    @Test
    fun `getAuthHeader when both are blank`() {
        every { parametersProvider.getBearerToken() } returns "temp"
        val client = ClassicJiraApiClient(parametersProvider, objectMapper)
        every { parametersProvider.getBearerToken() } returns ""
        every { parametersProvider.getBasicCredentials() } returns ":"
        assertNull(invokeGetAuthHeader(client))
    }

    private fun invokeGetAuthHeader(client: ClassicJiraApiClient): String? {
        val method = ClassicJiraApiClient::class.java.getDeclaredMethod("getAuthHeader")
        method.isAccessible = true
        try {
            return method.invoke(client) as? String
        } catch (e: java.lang.reflect.InvocationTargetException) {
            throw e.cause ?: e
        }
    }
}
